Making empty prefix list not match anything in juniper firewalls

haverma opened the pull request:
- Previously empty prefix list would cause all the following lines to be unreachable as it matched any packet

arifogel commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/representation/juniper/FwFromSourcePrefixList.java, line 39 at r1](https://reviewable.io/reviews/batfish/batfish/1367#-LCFQD4y0_zuzJ9VT5cV:-LCFQD4y0_zuzJ9VT5cW:bbs4u1o) ([raw file](https://github.com/batfish/batfish/blob/a5b257ed708fc0e20ad6463532bc83a02c715bbe/projects/batfish/src/main/java/org/batfish/representation/juniper/FwFromSourcePrefixList.java#L39)):*
> ```Java
>       // if referenced prefix list is empty, it should not match anything
>       if (sourcePrefixList.getLines().isEmpty()) {
>         headerSpaceBuilder.addSrcIp(EmptyIpSpace.INSTANCE);
> ```

Won't this cause NPE if prefix-list with that name does not exist? Applies elsewhere as well.

---

*[projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java, line 1203 at r1](https://reviewable.io/reviews/batfish/batfish/1367#-LCFRSWr2LNajEz-U7HI:-LCFRSWr2LNajEz-U7HJ:b-u0iqx9) ([raw file](https://github.com/batfish/batfish/blob/a5b257ed708fc0e20ad6463532bc83a02c715bbe/projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java#L1203)):*
> ```Java
> 
>     IpAccessList incomingFilter = c.getInterfaces().get("p-e.0").getIncomingFilter();
>     IpAccessList incomingFilterExcept = c.getInterfaces().get("p-e-except.0").getIncomingFilter();
> ```

I don't think you should grab the IpAccessList this way. Better to just grab via `c.getIpAccessLists()`. Otherwise you are relying on correct implementation of interface ACL assignment, which is not really the purpose of this test.
You should also remove the interfaces from the test config.

---

*[projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java, line 1210 at r1](https://reviewable.io/reviews/batfish/batfish/1367#-LCEzTuo1s1HYJB1ZuHB:-LCFRK157TUjIFBM6bMY:bib7yhk) ([raw file](https://github.com/batfish/batfish/blob/a5b257ed708fc0e20ad6463532bc83a02c715bbe/projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java#L1210)):*
<details><summary><i>Previously, sfraint (Spencer Fraint) wrote…</i></summary><blockquote>

Is this correct? or should it `accept` instead of `reject`?
I would expect the `filterExcept` behavior to be the opposite of the `filter` behavior.
</blockquote></details>

Yeah.. how is this test passing?

---


*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1367)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java, line 1228 at r2](https://reviewable.io/reviews/batfish/batfish/1367#-LCFVHMT0PTHrNipR-UG:-LCFVHMT0PTHrNipR-UH:b-1heqd9) ([raw file](https://github.com/batfish/batfish/blob/8f16b69b662d14fb4dd7a01ba177475aed83d349/projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java#L1228)):*
> ```Java
> 
>     // Every destination IP should match the empty prefix list
>     assertThat(incomingFilterDestinationxcept, accepts(testFlow1, "p-e-d-e.0", c));
> ```

missing `E` in `xcept`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1367)*
<!-- Sent from Reviewable.io -->

haverma commented:
Review status: 5 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[a discussion](https://reviewable.io/reviews/batfish/batfish/1367#-LCEy-1r1jXVmU__Lrlj:-LCFbpAhEOf_irZyovza:b-ld0t5u) (no related file):*
<details><summary><i>Previously, sfraint (Spencer Fraint) wrote…</i></summary><blockquote>

Is this also relevant to `FwFromDestinationPrefixList`, `FwFromDestinationPrefixListExcept`, and `FwFromPrefixList`?
</blockquote></details>

Yes, thanks, added changes in those as well

---

*[projects/batfish/src/main/java/org/batfish/representation/juniper/FwFromSourcePrefixList.java, line 39 at r1](https://reviewable.io/reviews/batfish/batfish/1367#-LCFQD4y0_zuzJ9VT5cV:-LCFaVu98cRRpllerIm5:bw1gb1p) ([raw file](https://github.com/batfish/batfish/blob/a5b257ed708fc0e20ad6463532bc83a02c715bbe/projects/batfish/src/main/java/org/batfish/representation/juniper/FwFromSourcePrefixList.java#L39)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Won't this cause NPE if prefix-list with that name does not exist? Applies elsewhere as well.
</blockquote></details>

We are checking `PrefixList pl !=null` while entering this function, won't this imply `RouteFilterList` by the same name will also be not null

---

*[projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java, line 1203 at r1](https://reviewable.io/reviews/batfish/batfish/1367#-LCFRSWr2LNajEz-U7HI:-LCFbDzmD-4vX4tpPuyM:b-65b3pm) ([raw file](https://github.com/batfish/batfish/blob/a5b257ed708fc0e20ad6463532bc83a02c715bbe/projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java#L1203)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

I don't think you should grab the IpAccessList this way. Better to just grab via `c.getIpAccessLists()`. Otherwise you are relying on correct implementation of interface ACL assignment, which is not really the purpose of this test.
You should also remove the interfaces from the test config.
</blockquote></details>

getting `IpAccessList` directly by its name 

---

*[projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java, line 1210 at r1](https://reviewable.io/reviews/batfish/batfish/1367#-LCEzTuo1s1HYJB1ZuHB:-LCFbNTq6fMPfKK0Eblq:b-9hkupv) ([raw file](https://github.com/batfish/batfish/blob/a5b257ed708fc0e20ad6463532bc83a02c715bbe/projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java#L1210)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Yeah.. how is this test passing?
</blockquote></details>

Yes, it should reject. Corrected the `FwFrom*Except` changes to comply to the behavior observed on Juniper VM router

---

*[projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java, line 1228 at r2](https://reviewable.io/reviews/batfish/batfish/1367#-LCFVHMT0PTHrNipR-UG:-LCFbmYP94y5TupXCRKn:b-yigp24) ([raw file](https://github.com/batfish/batfish/blob/8f16b69b662d14fb4dd7a01ba177475aed83d349/projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java#L1228)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

missing `E` in `xcept`.
</blockquote></details>

corrected

---


*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1367)*
<!-- Sent from Reviewable.io -->

haverma commented:
Review status: 5 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java, line 1210 at r1](https://reviewable.io/reviews/batfish/batfish/1367#-LCEzTuo1s1HYJB1ZuHB:-LCFcB349iGv71--PcTN:b19jxpo) ([raw file](https://github.com/batfish/batfish/blob/a5b257ed708fc0e20ad6463532bc83a02c715bbe/projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java#L1210)):*
<details><summary><i>Previously, haverma (Harsh Verma) wrote…</i></summary><blockquote>

Yes, it should reject. Corrected the `FwFrom*Except` changes to comply to the behavior observed on Juniper VM router
</blockquote></details>

Correction - it should "accept" all packets

---


*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1367)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/representation/juniper/FwFromSourcePrefixList.java, line 39 at r1](https://reviewable.io/reviews/batfish/batfish/1367#-LCFQD4y0_zuzJ9VT5cV:-LCFilQFB4gBy0Pdi5TX:b-p1mwqe) ([raw file](https://github.com/batfish/batfish/blob/a5b257ed708fc0e20ad6463532bc83a02c715bbe/projects/batfish/src/main/java/org/batfish/representation/juniper/FwFromSourcePrefixList.java#L39)):*
<details><summary><i>Previously, haverma (Harsh Verma) wrote…</i></summary><blockquote>

We are checking `PrefixList pl !=null` while entering this function, won't this imply `RouteFilterList` by the same name will also be not null
</blockquote></details>

Sorry, I should have viewed more context first.
This is fine as is.

---

*[projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java, line 1212 at r3](https://reviewable.io/reviews/batfish/batfish/1367#-LCFiRNx3b32Z6L6bUM1:-LCFiRNy7D5bQOcAY0P6:bn7cyt9) ([raw file](https://github.com/batfish/batfish/blob/f831ef9fd374e194ad8673ba18a2d1196d9aa2c7/projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java#L1212)):*
> ```Java
> 
>     // No source IP should match the empty prefix list
>     assertThat(incomingFilterSource, rejects(testFlow1, "interface", c));
> ```

You don't need to specify a `srcInterface`.  `null` is perfectly acceptable - it just means that the flow originated from the device, instead of being received on some interface.

---


*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1367)*
<!-- Sent from Reviewable.io -->

haverma commented:
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java, line 1212 at r3](https://reviewable.io/reviews/batfish/batfish/1367#-LCFiRNx3b32Z6L6bUM1:-LCG6mcZDfONSMCGJPyx:b-pbzxiw) ([raw file](https://github.com/batfish/batfish/blob/f831ef9fd374e194ad8673ba18a2d1196d9aa2c7/projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java#L1212)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

You don't need to specify a `srcInterface`.  `null` is perfectly acceptable - it just means that the flow originated from the device, instead of being received on some interface.
</blockquote></details>

replaced with `null`

---


*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1367)*
<!-- Sent from Reviewable.io -->

sfraint commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 4 files at r1, 4 of 6 files at r2, 1 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1367#-:-LCUQeEV0jglXPWf8Ncy:bnfp4nl)*
<!-- Sent from Reviewable.io -->
